diff --git a/third_party/ffmpeg/chromium/scripts/build_ffmpeg.py b/third_party/ffmpeg/chromium/scripts/build_ffmpeg.py
index f969db3..f941018 100755
--- a/third_party/ffmpeg/chromium/scripts/build_ffmpeg.py
+++ b/third_party/ffmpeg/chromium/scripts/build_ffmpeg.py
@@ -246,12 +246,11 @@ def main(argv):
       '--disable-xlib',
       '--disable-zlib',
 
-      # Disable hardware decoding options which will sometimes turn on
-      # via autodetect.
+      # Enable hardware decoding options
       '--disable-dxva2',
-      '--disable-vaapi',
-      '--disable-vda',
-      '--disable-vdpau',
+      '--enable-vaapi',
+      '--enable-vda',
+      '--enable-vdpau',
 
       # Common codecs.
       '--enable-decoder=theora,vorbis,vp8',
